### PR TITLE
PyTorch 1.7 Stable support

### PR DESCRIPTION
**Pull request by ydcjeff**
## What does this PR do?

Conda 1.7 test is failing again and found out PyTorch 1.7 is in RC stage and 1.7 RC binary cannot be found on conda.

This PR
- prepares for PyTorch 1.7 stable once it's released
- **doesn't add PyTorch 1.7 new features**
- will fix the failed tests for 1.7 Stable where applicable

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
in meantime, we can drop it from CI #3833 or add a special install task as it is available on `pip --pre` as pre-release

**Comment by ydcjeff** [comment]
Is RC version available on pypi? I didn't find it.
PyTorch released RC version via only github release

**Comment by Borda** [comment]
@ydcjeff can we now install PT 1.7 from regular channels - PyPI and Conda?

**Comment by ydcjeff** [comment]
> @ydcjeff can we now install PT 1.7 from regular channels - PyPI and Conda?

I think not yet

**Comment by ydcjeff** [comment]
@Borda There is no 1.7 docker on docker hub.

**Comment by ydcjeff** [comment]
Pyright and conda 1.7 checks are expected to fail.
The rest aren't.
DDP support on windows is prototype feature, so currently skipping it.

**Review by awaelchli — APPROVED** [review]
nice, is it ready?

**Review by SeanNaren — APPROVED** [review]

File: tests/checkpointing/test_model_checkpoint.py [hunk 1]
```diff
@@ -367,7 +367,7 @@ def test_model_checkpoint_topk_all(tmpdir):
     seed_everything(1000)
     epochs = 2
     model = EvalModelTemplate()
-    checkpoint_callback = ModelCheckpoint(dirpath=tmpdir, monitor="early_stop_on", save_top_k=-1)
+    checkpoint_callback = ModelCheckpoint(dirpath=tmpdir, monitor="some_val", save_top_k=-1)
```
Comment (tchaton) [inline tests/checkpointing/test_model_checkpoint.py:370]: Why did you modify some tests ?

**Review by tchaton — CHANGES_REQUESTED** [review]
Great catch ! I was wondering why some tests were modified ?

**Review by tchaton — APPROVED** [review]

File: tests/checkpointing/test_model_checkpoint.py (hunk 1 above)
Comment (ydcjeff) [reply tests/checkpointing/test_model_checkpoint.py:370]: @tchaton  That test was failing on macos and windows (often) with PyTorch 1.7
but it is passing locally with 1.7. Tried with some other monitor and test god passed.
So requested for a review for a while.

**Comment by ydcjeff** [comment]
> nice, is it ready?

I think so, currently model_checkpoint test passed with some other monitor, not with `early_stop_on`
grad norm also has to round to 4 points to pass. but I am not convinced about those.

File: tests/checkpointing/test_model_checkpoint.py (hunk 1 above)
Comment (awaelchli) [reply tests/checkpointing/test_model_checkpoint.py:370]: I will have a look at that test but it makes sense to me this test starts failing, because the last checkpoint is not guaranteed to be the best. the monitor key should correspond to the training- or validation loss, not early stopping.

Comment (ydcjeff) [reply tests/checkpointing/test_model_checkpoint.py:370]: But early_stop_on is defined in validation_epoch_end

**Review by SkafteNicki — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: tests/checkpointing/test_model_checkpoint.py (hunk 1 above)
Comment (ydcjeff) [reply tests/checkpointing/test_model_checkpoint.py]: Ahh, tests are producing different best_model_path
https://github.com/PyTorchLightning/pytorch-lightning/actions/runs/333879691

Comment (awaelchli) [reply tests/checkpointing/test_model_checkpoint.py]: yes it seems the change in pytorch version produces slightly different results which leads to different best model path... it's a bit strange.

Comment (SeanNaren) [reply tests/checkpointing/test_model_checkpoint.py]: Any findings/updates here @ydcjeff? Don't mind assisting here to find the source of this issue (we probably shouldn't merge till we know why this happens)

Comment (ydcjeff) [reply tests/checkpointing/test_model_checkpoint.py]: Thank you @SeanNaren 
[this](https://github.com/PyTorchLightning/pytorch-lightning/pull/3821/checks?check_run_id=1327084401) and [this result](https://github.com/PyTorchLightning/pytorch-lightning/pull/3821/checks?check_run_id=1327084438) are different although we are doing same seed with different pytorch versions.

The comparison is early_stop_on values.
How should I approach this? The ModelCheckpoint is saving the best model tho.

**Comment by ydcjeff** [comment]
@awaelchli @tchaton @SeanNaren 
ready for final review.

File: tests/checkpointing/test_model_checkpoint.py (hunk 1 above)
Comment (SeanNaren) [reply tests/checkpointing/test_model_checkpoint.py]: @ydcjeff Were you able to reproduce this locally? I wasn't able to...

Comment (ydcjeff) [reply tests/checkpointing/test_model_checkpoint.py]: No, I can't reproduce the CI errors. I also have one machine.
The results are always same on same machines, different on different machines and pytorch versions.
And that's normal
See: https://discuss.pytorch.org/t/reproducibility-over-different-machines/63047/2

Comment (awaelchli) [reply tests/checkpointing/test_model_checkpoint.py]: I have windows here and PT 1.7 and the test runs, it must have something todo with the platform and the random number generator.

@ydcjeff increased the epochs in a previous commit https://github.com/PyTorchLightning/pytorch-lightning/pull/3821/commits/9c649c725a376037464c3fabfe1ce38cf47a861d
and it showed the best model according to the validation loss was not the last, which would mean this example is either overfitting or not learning at all, leading to some random checkpoint to be the best.

File: tests/checkpointing/test_model_checkpoint.py [hunk 2]
```diff
@@ -365,20 +365,24 @@ def test_model_checkpoint_topk_zero(tmpdir):
 def test_model_checkpoint_topk_all(tmpdir):
     """ Test that save_top_k=-1 tracks the best models when monitor key is provided. """
```
Comment (awaelchli) [inline tests/checkpointing/test_model_checkpoint.py:366]: I was hoping this test could be more explicit, showing that the last is the best and the first is the worst, but I guess it's not so easy with the randomness. 
I will look into this more

**Review by awaelchli — APPROVED** [review]

File: tests/checkpointing/test_model_checkpoint.py (hunk 1 above)
Comment (ydcjeff) [reply tests/checkpointing/test_model_checkpoint.py]: Yep, the model is kinda overfitting, the only way I can think of now is tracking with `monitor`

File: tests/checkpointing/test_model_checkpoint.py (hunk 2 above)
Comment (rohitgr7) [reply tests/checkpointing/test_model_checkpoint.py:366]: what if monitor is set to `epoch` with `mode='max'`? this way we can ensure first is the worst and last is the best.

Comment (ydcjeff) [reply tests/checkpointing/test_model_checkpoint.py:366]: But here is we need to track the best model, isn't it? not smallest or largest epoch, I think?

Comment (rohitgr7) [reply tests/checkpointing/test_model_checkpoint.py:366]: yep, it's not ideally correct to consider largest epoch as the best one but since I have set my `monitor='epoch'` with `mode='max'` in such a case checkpoint should monitor `epoch` only and at the end largest epoch will be the best one for the checkpoint_callback.

Comment (awaelchli) [reply tests/checkpointing/test_model_checkpoint.py:366]: hey, @rohitgr7 I like this idea. This is a valid test because for it to work properly it needs to track the metric correctly, and that is what we want to test! Can we try it?

Comment (ydcjeff) [reply tests/checkpointing/test_model_checkpoint.py:366]: I see, I track smallest `early_stop_on` here, both are valid.
not sure what others think about this?

Comment (ydcjeff) [reply tests/checkpointing/test_model_checkpoint.py:366]: Okay, monitor `epoch` is not in the return metrics

Comment (rohitgr7) [reply tests/checkpointing/test_model_checkpoint.py:366]: I think it's logged by default. If it isn't.. yeah you can subclass `EvalTemplateModel`, override `validation_epoch_end` and log it in the test itself.

Comment (awaelchli) [reply tests/checkpointing/test_model_checkpoint.py:366]: if it's too many changes, I'd rather keep it for now and follow up with a PR that improves the test. 
I think this PR is high priority to restore master branch from failing.

**Comment by ydcjeff** [comment]
Once the CI passed, if ok with the changes, ready to merge.

**Comment by SeanNaren** [comment]
There is a failing test which seems to not find a pytorch 1.7 base docker image. Do we need to manually create this?

**Comment by ydcjeff** [comment]
> There is a failing test which seems to not find a pytorch 1.7 base docker image. Do we need to manually create this?

@SeanNaren we create docker image nightly and after this is merged, that will be created as well.
But yes, if we wish to create and push to docker hub, will wait for that.

File: tests/checkpointing/test_model_checkpoint.py (hunk 2 above)
Comment (ydcjeff) [reply tests/checkpointing/test_model_checkpoint.py:366]: Changed!

**Review by rohitgr7 — APPROVED** [review]

File: tests/checkpointing/test_model_checkpoint.py (hunk 2 above)
Comment (awaelchli) [reply tests/checkpointing/test_model_checkpoint.py:366]: haha awesome!

**Comment by ydcjeff** [comment]
pinning @SeanNaren for final review.

**Comment by SeanNaren** [comment]
thanks @ydcjeff, will monitor to see if everything goes smoothly with the ci docker workflows!

**Comment by SeanNaren** [comment]
hey @ydcjeff do we need to add 1.7 to the docker build push here? https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/workflows/docker-builds.yml#L17

**Comment by ydcjeff** [comment]
> hey @ydcjeff do we need to add 1.7 to the docker build push here? https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/workflows/docker-builds.yml#L17

no need currently, this is pytorch lightning docker image to be used by our users
this is not for CI testing.

**Comment by SeanNaren** [comment]
Are we missing the public pytorch 1.7 docker image for this test? https://github.com/PyTorchLightning/pytorch-lightning/runs/1332900421

Any idea what's happening here? https://github.com/PyTorchLightning/pytorch-lightning/runs/1332770138

**Comment by ydcjeff** [comment]
> Are we missing the public pytorch 1.7 docker image for this test? https://github.com/PyTorchLightning/pytorch-lightning/runs/1332900421

Yes, the image will be built at UTC 0 nightly, so the tests will run tmr.

> Any idea what's happening here? https://github.com/PyTorchLightning/pytorch-lightning/runs/1332770138

No idea, that probably comes from master before this PR.
